utils/spdx/src/main/kotlin/parser/SpdxExpressionParser.kt

```diff
@@ -0,0 +1,196 @@
+/*
+ * Copyright (C) 2024 The ORT Project Authors (see <https://github.com/oss-review-toolkit/ort/blob/main/NOTICE>)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package org.ossreviewtoolkit.utils.spdx.parser
+
+import org.ossreviewtoolkit.utils.common.nextOrNull
+import org.ossreviewtoolkit.utils.spdx.SpdxCompoundExpression
+import org.ossreviewtoolkit.utils.spdx.SpdxExpression
+import org.ossreviewtoolkit.utils.spdx.SpdxLicenseIdExpression
+import org.ossreviewtoolkit.utils.spdx.SpdxLicenseReferenceExpression
+import org.ossreviewtoolkit.utils.spdx.SpdxLicenseWithExceptionExpression
+import org.ossreviewtoolkit.utils.spdx.SpdxOperator
+
+/**
+ * A parser for SPDX expressions. It consumes a sequence of [Token]s and produces an [SpdxExpression].
+ *
+ * This parser implements the grammar defined in the
+ * [SPDX specification](https://spdx.github.io/spdx-spec/v2.2.2/SPDX-license-expressions/):
+ *
+ * ```
+ * license-expression   -> simple-expression | compound-expression
+ * compound-expression  -> simple-expression |
+ *                         simple-expression "WITH" license-exception-id |
+ *                         compound-expression "AND" compound-expression |
+ *                         compound-expression "OR" compound-expression |
+ *                         "(" compound-expression ")" )
+ * simple-expression    -> license-id | license-id"+" | license-ref
+ * license-ref          -> ["DocumentRef-" idstring ":"] "LicenseRef-" idstring
+ * license-exception-id -> <short form license exception identifier in Annex A.2>
+ * license-id           -> <short form license identifier in Annex A.1>
+ * idstring             -> 1*(ALPHA / DIGIT / "-" / "." )
+ * ```
+ *
+ * For more efficient parsing, this is transformed into the following form which implements the operator precedence as
```

Comment:
Would you know why this approach is more efficient? (Would it be worth mentioning here briefly)

Reply:
I have reworded the sentence. It's not about efficient parsing, but about having a grammar that is easy to implement. Having the operator precedence as part of the grammar makes it easy that each line in the grammar matches a function in the parser and no additional logic is required to handle the operator precedence. The original grammar does not say anything about operator precedence, it's only explained in the documentation. Most examples I found for expression parsers use a grammar like this.